### PR TITLE
Add support for Serial/TTY

### PR DESCRIPTION
**Pull request by x86fr**

### CODE REVIEW

File: app/config.c [hunk 1]
```diff
@@ -129,6 +158,14 @@ static void parse_option(const char *option, const char *params)
         } else if (strncmp(params, "high", 5) == 0) {
             power_save = POWER_SAVE_HIGH;
         }
+    } else if (strncmp(option, "console", 8) == 0) {
+        if (params != NULL) {
+            parse_serial_params(params);
+        }
+    } else if (strncmp(option, "nobench", 8) == 0) {
```
Comment (debrouxl) [inline app/config.c:165]: Do we want to keep alphabetical ordering ? In which case the previous location for the nopause and nobench flag ordering was better.

File: app/config.c [hunk 2]
```diff
@@ -321,9 +360,17 @@ static void test_selection_menu(void)
         display_enabled(POP_R+12, i, test_list[i].enabled);
     }
 
+    bool tty_update = true;
     bool exit_menu = false;
     while (!exit_menu) {
         bool changed = false;
+
+        if (enable_tty && tty_update) {
```
Comment (debrouxl) [inline app/config.c:368]: This block is repeated multiple times across this file, so maybe there should be a macro for that ?

File: app/display.c [hunk 3]
```diff
@@ -175,6 +184,10 @@ void display_start_run(void)
         next_spin_time = run_start_time + SPINNER_PERIOD * clks_per_msec;
     }
     display_spinner('-');
+
+    if(enable_tty){
```
Comment (debrouxl) [inline app/display.c:188]: Whitespace :)

File: app/error.c [hunk 4]
```diff
@@ -365,5 +367,9 @@ void error_update(void)
                                    test_list[test_num].errors);
         }
         display_error_count(error_count);
+
+        if(enable_tty) {
```
Comment (debrouxl) [inline app/error.c:371]: Whitespace :)

File: lib/string.c [hunk 5]
```diff
@@ -105,3 +120,29 @@ char *strstr(const char *haystack, const char *needle)
...
+        num /= 10;
+    }
+
+    // Last is string terminator
+    str[i] = '\0';
+
+    reverse(str);
+
+    return str;
+}
```
Comment (debrouxl) [inline lib/string.c:148]: Newline at EOF.

File: system/keyboard.c [hunk 6]
```diff
@@ -255,5 +258,13 @@ char get_key(void)
         }
     }
 
+    if (enable_tty) {
+        uint8_t c = tty_get_key();
+        if (c != 0xFF) {
+            if(c == 0x0D) c = '\n'; // Enter
```
Comment (debrouxl) [inline system/keyboard.c:264]: Whitespace.

File: system/serial.c [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	if (!port->enable) {
+		return;
+	}
+	/* Now, do each character */
+	while (*p) {
+		serial_wait_for_xmit(port);
+
+		/* Send the character out. */
+		serial_write_reg(port, UART_TX, *p);
+		if(*p==10) {
```
Comment (debrouxl) [inline system/serial.c:76]: Whitespace.

File: system/serial.h [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+#define SERIAL_PORT_0x2F8   1
+#define SERIAL_PORT_0x3E8   2
+#define SERIAL_PORT_0x2E8   3
+
+static const uint16_t serial_base_ports[] = { 0x3f8, 0x2f8, 0x3e8, 0x2e8 };
+
+struct serial_port {
+	bool enable;
+	bool is_mmio;
+	unsigned int parity;
```
Comment (debrouxl) [inline system/serial.h:28]: Could use narrower data types for these, but it wouldn't have much of an impact on the code size.

File: system/serial.c [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	if (!enable_tty) {
+	    return;
+	}
+
+	int uart_status, serial_div;
+	unsigned char lcr;
+
+    console_serial.enable       = true;
+    console_serial.is_mmio      = false;
+    console_serial.baudrate     = tty_params_baud;
```
Comment (debrouxl) [inline system/serial.c:114]: The order of declaration and order of initialization do not match.

File: system/serial.c [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+            p[28] = '\0';
+            tty_print(row,0,p);
+            tty_normal();
+            p[28] = '|';
+            tty_print(row, 28, p + 28);
+            continue;
+        }
+
+        // Replace degree symbol with '*' for tty to avoid a C437/VT100 translation table.
+        if(row == 7 && col_len > 77 && (shadow_buffer[7][73].value & 0xFF) == 0xF8) {
```
Comment (debrouxl) [inline system/serial.c:187]: Whitespace.

File: system/serial.c [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+            tty_normal();
+        }
+    }
+}
+
+char tty_get_key(void)
+{
+    int uart_status = serial_read_reg(&console_serial, UART_LSR);
+
+	if (uart_status & UART_LSR_DR) {
```
Comment (debrouxl) [inline system/serial.c:205]: Whitespace.

File: system/serial.h [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+#define TTY_UNDERLINE       "\x1b[4m"
+#define TTY_INVERSE         "\x1b[7m"
+
+/*
+ * Definitions for the Base UART Registers
+ */
+
+#define UART_REF_CLK 1843200
+
+#define UART_RX		0	/* In:  Receive buffer (DLAB=0) */
```
Comment (debrouxl) [inline system/serial.h:55]: Whitespace.

File: system/serial.h [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+#define UART_LCR_WLEN6  0x01	/* Wordlength: 6 bits */
+#define UART_LCR_WLEN7  0x02	/* Wordlength: 7 bits */
+#define UART_LCR_WLEN8  0x03	/* Wordlength: 8 bits */
+
+/*
+ * Definitions for the Line Status Register
+ */
+#define UART_LSR_TEMT	0x40	/* Transmitter empty */
+#define UART_LSR_THRE	0x20	/* Transmit-hold-register empty */
+#define UART_LSR_BI     0x10	/* Break interrupt indicator */
```
Comment (debrouxl) [inline system/serial.h:90]: Whitespace here and in the following code blocks.

Comment (x86fr) [reply system/serial.h:90]: PS: Most of these blank lines are intentional. They basically increase readability by separating code blocks (or isolate important line visually). Removing all blank lines add a lot of obfuscation to the code.

Comment (debrouxl) [reply system/serial.h:90]: The blank lines are fine indeed, but there's inconsistent indentation of the right side of #defines :)

File: system/keyboard.c (hunk 6 above)
Comment (x86fr) [reply system/keyboard.c:264]: I'm doing a pass to solve all tabs->spaces, but I can't see the issue here.

Comment (debrouxl) [reply system/keyboard.c:264]: Most other places use `if (` :)

**Comment by debrouxl** [comment]
Hmm... I seem to have failed to send a comment about rebasing the branch first, instead of using the Rebase and merge Github option, in order to move conflict resolution from a merge commit to the commits where the changes belong.

In my experience at work, conflict resolution as part of a merge commit proves to be a pain when history rewrites (usually to be avoided) are needed for some reason: they break the neat automatic history linearization `git rebase --interactive` can usually perform with these conflict-resoluting merge commits.

File: system/keyboard.c (hunk 6 above)
Comment (x86fr) [reply system/keyboard.c:264]: ah yes, I was looking for superfluous spaces :) 

I always had trouble switching between coding convention. Ie: many devs I work with consider anything other than a 2-spaces indent as heretic (but they also loves to write obscure one liner...) .

**Comment by x86fr** [comment]
For this specific PR, I plan to squash it as it's a single feature.

**Comment by debrouxl** [comment]
Oh yeah, squashing is a pretty fine - and simpler, besides - way to handle this PR :)

File: app/config.c (hunk 2 above)
Comment (martinwhitaker) [reply app/config.c:368]: Alternatively, change all occurrences of `tty_update = true` to `tty_update = enable_tty`. then you can simply use `if (tty_update)`.

File: app/config.c [hunk 14]
```diff
@@ -321,9 +360,17 @@ static void test_selection_menu(void)
...
+            tty_popup_redraw();
+            tty_update = false;
```
Comment (martinwhitaker) [inline app/config.c:370]: This is redundant, because you immediately set tty_update back to true (same in other places).

File: app/display.c [hunk 15]
```diff
@@ -69,14 +73,16 @@ void display_init(void)
 
     clear_screen();
 
-    set_foreground_colour(RED);
-    set_background_colour(WHITE);
+    set_foreground_colour(BLACK);
+    set_background_colour(GREEN);
```
Comment (martinwhitaker) [inline app/display.c:77]: This is unrelated to serial/TTY support, so belongs in a separate commit.

File: app/config.c [hunk 16]
```diff
@@ -44,27 +45,6 @@
...
-#define POP_REGION  POP_R, POP_C, POP_LAST_R, POP_LAST_C
-
-#define POP_LM      (POP_C + 3)     // Left margin
-#define POP_LI      (POP_C + 5)     // List indent
-
-#define SEL_W       32
-#define SEL_H       2
-
-#define SEL_AREA    (SEL_W * SEL_H)
-
```
Comment (martinwhitaker) [inline app/config.c:67]: You could avoid making all these constants public by using `tty_send_region(POP_REGION)` instead of `tty_popup_redraw`. I think that would be just as readable.

File: system/screen.h [hunk 17]
```diff
@@ -19,6 +19,26 @@
 #define SCREEN_WIDTH    80
 #define SCREEN_HEIGHT   25
 
+
+typedef struct {
+    uint8_t     r;
+    uint8_t     g;
+    uint8_t     b;
+} __attribute__((packed)) rgb_value_t;
```
Comment (martinwhitaker) [inline system/screen.h:27]: I don't think you need to make this type public.

File: app/config.c (hunk 14 above)
Comment (x86fr) [reply app/config.c:370]: yes! the redraw block used to be at the end of the function :-/

File: app/display.c (hunk 15 above)
Comment (x86fr) [reply app/display.c:77]: Correct. That belong to the aesthetic branch (and also the red + some line below). Anyway, I'll try to push that quickly to main because the actual color confuses some betatesters (I still get bug reports for passmark's memtest86 :-/). Removing here.

File: system/keyboard.c (hunk 6 above)
Comment (martinwhitaker) [reply system/keyboard.c]: I'm quite used to switching. For my own code I use Camel case and an indentation of 2. I also help maintain Icarus Verilog - the only code base I know of that uses an indentation of 6.

File: app/display.c (hunk 15 above)
Comment (martinwhitaker) [reply app/display.c:77]: I find black on green hard to read which is why I changed the colour scheme.

File: app/config.c (hunk 16 above)
Comment (x86fr) [reply app/config.c:67]: I moved them for the popup redraw in read.c, but I will revert that abd only redraw the row/col region passed to `read_value()` instead in that loop

File: app/config.c (hunk 2 above)
Comment (x86fr) [reply app/config.c:368]: Yes!

File: app/display.c (hunk 15 above)
Comment (x86fr) [reply app/display.c:77]: I also have the same concern about the RED/WHITE color scheme :-/ 

Anyway, I don't have any fetish for the green color. It was useful for a very long time because the original memtest86 was red and memtest86+ was green, so anyone could tell at a glance (even from a distance) if a system was running memtest86 or memtest86+. 

But now, passmark changed the title for their memtest86 from red to green so that didn't work anymore.

I'll do some tries next week and ask for feedbacks
